modules/dnn/test/test_onnx_importer.cpp

```diff
@@ -2066,7 +2066,7 @@ TEST_P(Test_ONNX_nets, Googlenet)
         applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_NGRAPH);
 #endif
 
-    const String model = _tf("models/googlenet.onnx", false);
```

Comment:
We should not introduce regressions for previously supported models.

You could add new test cases, but you can't just drop them.

Reply:
It is just wrong when you force it running with batch size >= 2 while it has a reshape operator always reshaping the batch size to be 1 (more details [here](https://github.com/opencv/opencv_extra/pull/1073#issue-1782223950). Why do we have to keep a wrong thing in our tests? It does not make sense to replicate bugs in a refactor implementation.

Reply:
It is just wrong when you force it running with batch size >= 2 while it has a reshape operator always reshaping the batch size to be 1 (more details [here](https://github.com/opencv/opencv_extra/pull/1073#issue-1782223950). Why do we have to keep a wrong thing in our tests? It does not make sense to replicate bugs in a refactor implementation.

Reply:
@dkurt Could you please take a look?

Reply:
Am I right that original model is not available anymore? https://s3.amazonaws.com/download.onnx/models/opset_8/bvlc_googlenet.tar.gz

```
Model <GoogleNet (ONNX)>
  expect 739732220ba2e3efa88f7c26f13badad9b7514bc
  catch [Errno 2] No such file or directory: 'bvlc_googlenet.tar.gz'
  hash check failed - downloading
  get https://s3.amazonaws.com/download.onnx/models/opset_8/bvlc_googlenet.tar.gz
  catch HTTP Error 403: Forbidden
```

Also, if a new model from https://drive.google.com/u/0/uc?id=1FucNLURGgdPk4nCxT0378isPigCcbZ1M&export=download is a personal Google Drive with a modified file, I think it's not a good idea.

Reply:
Due even opset 12 version has Reshape with 1x1024 hardcoded shapes, I think it does not make sense to update the model with Flatten. However still there is a need to update URL/model/data. So I recommend this: https://github.com/onnx/models/tree/main/vision/classification/inception_and_googlenet/googlenet

Reply:
> Also, if a new model is a personal Google Drive with a modified file

I dont find a proper place to store the model. Do we have a shared cloud drive for this? IIRC, Maksim worked on this previously. Let us include him @mshabunin for this topic.

> Due even opset 12 version has Reshape with 1x1024 hardcoded shapes

It has hardcoded 1x1024 Reshape because the input shape has batch size 1. The ONNX Gemm operator does not support batch size. Although technically speaking we can make it support batch size, it will only cause more troubles like this in the future (and in other backends) if we do so. I suggest we just fail and exit the program if the model proto is defined to have batch size 1 and has some operators does not support batch size like Gemm but yet the user still want to run it with batch size >= 2.

Reply:
Why not to use GEMM implementation inside InnerProduct in case `batch_size=1`? Like `fastConvImpl` from Convolution layer? Call GEMM when it's supported and InnerProduct if not.

https://github.com/opencv/opencv/blob/1794cdc03c9505bb46f33a5cde5e210c1c7f65a4/modules/dnn/src/layers/convolution_layer.cpp#L260

Reply:
The idea was to use this refactor to replace InnerProduct in the end. Code in InnerProduct is a mess and not maintainable.

Reply:
@fengyuentau, good point, but there are the following issues:
* No backends support in GEMM. 
* We need to replace InnerProduct to GEMM everywhere, not only ONNX
> | fc (current impl) | mean=21.94   median=21.40   min=19.74 |
* Which implementation of FC layer has been benchmarked? Does GEMM shows better performance on other CPUs? https://github.com/opencv/opencv/blob/1794cdc03c9505bb46f33a5cde5e210c1c7f65a4/modules/dnn/src/layers/fully_connected_layer.cpp#L268-L292


So my proposal is to determine for which configurations GEMM is better first.

Reply:
As PR is ready, let's discuss one more time why replacement of the official ONNX model is required? If possible, let's keep the origin one.

Reply:
Details in https://github.com/opencv/opencv_extra/pull/1073. Short version: the original one has a `Reshape` operator which forces batch size to be 1, while in the test we want it to run with batch size 2. This is just wrong.

Reply:
OpenCV is able to handle this. Adding the following lines at `getMemoryShapes` of ReshapeLayer reports correct shapes:
```cpp
std::cout << name << std::endl;
std::cout << inputs[0] << std::endl;
std::cout << outputs[0] << std::endl;
```
```
onnx_node_output_0!OC2_DUMMY_0
[ 2 1024 1 1 ]
[ 2 1 1024 ]
```

So you probably need to check ability of a Gemm layer to change the shape.

Reply:
Thanks, I will try to solve this issue in a different PR.

Reply:
Hold on. Although I am not fully agree all the opinions here, I am working on adding back this feature, which needs only two additional transposition.

Reply:
@fengyuentau, sure. Here is a brief proposal so if you find it suitable, we can add a workaround: https://github.com/dkurt/opencv/commit/cad21a3d0ab2b42c3d5d7d67b55606e8982e311f

Reply:
Man, I found if batch is allowed here, it is even much more broken. I found in `TEST_P(Test_ONNX_nets, Googlenet)`, the shape of `ref` is `2x1000` instead of `2x1x1000`, meaning the shape of output of the last Gemm layer in Googlenet is `2x1000` also. Note that the input of this Gemm layer is `2 x 1 x 1024` (2 is the batch added by `blobsFromImages` in the source input). This is a broken shape inference with batch in the test. 

Also I can see The forced-batch feature is going to cause some troubles in vision transformer models. Following is an example.

<img width="1333" alt="image" src="https://github.com/opencv/opencv/assets/17219438/0745fd59-6910-4a93-9c37-5173cabf43c7">

Above is the an attention block in ONNX. The batch dimension can be blended into other dimensions; In the 2nd last Reshape, the original batch size `1` is blended into `8`. If the forced batch feature brings a `batch=2`, the Gemm layer won't work because the inner dimension will not match.

Reply:
@fengyuentau, will https://github.com/dkurt/opencv/commit/cad21a3d0ab2b42c3d5d7d67b55606e8982e311f solve the problem at least partially?

Reply:
It should fix the googlenet problem but it messes the dimensions in attention block. `Transpose` operators in the attention block transposes the batch dimension to axis 1.

Could we just simply follow the standards instead of going futher in a wrong way? The forced-batch feature assumes the batch dimension is at axis 0, which may work for most of the conv nets but not for vit.